File: java/3sum.java
/**
*/
import java.util.ArrayList;
import java.util.Collections;

public class Solution {
   public ArrayList<ArrayList<Integer>> threeSum(int[] num) {
        // Start typing your Java solution below
        // DO NOT write main() function

        ArrayList<Integer> numArr = new ArrayList<Integer>();
        for(int i=0;i<num.length;i++){
            numArr.add(num[i]);
        }
        Collections.sort(numArr);
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer> >();
        for(int i=0;i<numArr.size();i++){
            if(numArr.get(i)>0) break;
            int j,k;
            j = i+1;
            k = numArr.size()-1;
            while(j<k){
                int tmpSum = numArr.get(i)+numArr.get(j)+numArr.get(k);
                if(tmpSum<0){
                    j++;

                }else if(tmpSum>0){
                    k--;
                }else{
                    ArrayList<Integer> tmpArray = new ArrayList<Integer>();

                    tmpArray.add(numArr.get(i));
                    tmpArray.add(numArr.get(j));
                    tmpArray.add(numArr.get(k));
                    result.add(tmpArray);


                    j++;
                    k--;
                }
            }
        }
        HashSet<ArrayList<Integer>> tmpset = new HashSet<ArrayList<Integer>>(result);
        result = new ArrayList<ArrayList<Integer>>(tmpset);

        return result;
    }

}

File: cpp/3Sum.cpp
class Solution {
public:
    vector<vector<int> > threeSum(vector<int> &num) {
        set<vector<int> > res;
        //vector<int> localSum(num);
        sort(num.begin(), num.end());
        for(int i=0;i+2<num.size();i++) {
            if(i>=1 && num[i] == num[i-1]) {
                continue;
            }
            int j=i+1;
            int k=num.size()-1;
            while(j<k) {
                int s = num[i]+num[j]+num[k];
                if(s==0) {
                    vector<int> tmp;
                    tmp.push_back(num[i]);
                    tmp.push_back(num[j]);
                    tmp.push_back(num[k]);
                    res.insert(tmp);
                    j++;
                    k--;
                }
                else if(s<0) {
                    j++;
                } else {
                    k--;
                }
            }
        }
        return vector<vector<int> >(res.begin(), res.end());
    }
};
